libc/stdio/stdio_indicator.c

```diff
@@ -0,0 +1,628 @@
+
+	for (int i = 0; i < num_modes; ++i) {
+		f = fopen(filename, modes[i]);
+		TEST_ASSERT_NOT_NULL_MESSAGE(f, tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL(-1, fgetc(f));
+		CHECK_MATCH(tmp, modes[i], "w", "a", "wb", "ab");
+		TEST_ASSERT_EQUAL_MESSAGE(tmp ? 0 : 1, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+	}
+	TEST_ASSERT_EQUAL(0, remove(filename));
+}
+
+
+TEST(stdio_feof, preserve_errno_huge_size)
+{
+	int errnoBefore;
+	int multiplier = 10;
+	long pos = 0;
+	const char *filename = "test_stdio_feof_errno_preserve";
+	FILE *f = fopen(filename, "w+");
+	TEST_ASSERT_NOT_NULL(f);
+
+	for (int i = 0; i < multiplier; i++) {
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fputs(testdata_hugeStr, f));
+	}
+
+	fputc('A', f);
+	errno = 0;
+	while (pos <= multiplier * testdata_hugeSize) {
+		TEST_ASSERT_EQUAL_INT(0, fseek(f, pos, SEEK_SET));
+		errnoBefore = errno;
+		TEST_ASSERT_EQUAL_INT(0, feof(f));
+		TEST_ASSERT_EQUAL(errnoBefore, errno);
+		pos += testdata_hugeSize / 50;
+		errno++;
+	}
+	TEST_ASSERT_EQUAL_INT(-1, fgetc(f));
+	errnoBefore = errno;
+	TEST_ASSERT_EQUAL_INT(1, feof(f));
+	TEST_ASSERT_EQUAL_INT(errnoBefore, errno);
+	fclose(f);
+	remove(filename);
+}
+
+
+TEST_GROUP_RUNNER(stdio_feof)
+{
+	RUN_TEST_CASE(stdio_feof, not_empty_all_modes);
+	RUN_TEST_CASE(stdio_feof, empty_all_modes);
+	RUN_TEST_CASE(stdio_feof, preserve_errno_huge_size)
+}
+
+
+TEST_GROUP(stdio_ftell);
+
+TEST_SETUP(stdio_ftell)
+{
+}
+
+
+TEST_TEAR_DOWN(stdio_ftell)
+{
+}
+
+
+TEST(stdio_ftell, correct_position_not_empty)
+{
+	FILE *f;
+	const char *filename = "test_stdio_ftell_not_empty";
+	char *data = testdata_createCharStr(ALL_ASCII);
+	int tmp;
+
+	for (int i = 0; i < num_modes; ++i) {
+		f = fopen(filename, "w+");
+		TEST_ASSERT_NOT_NULL(f);
+		TEST_ASSERT_NOT_NULL(data);
+		TEST_ASSERT_NOT_EQUAL_INT_MESSAGE(EOF, fputs(data, f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT32_MESSAGE(-1, (int32_t)ftell(f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT32_MESSAGE((int32_t)ftell(f), (int32_t)ftello(f), tellMode(modes[i]));
+
+		f = fopen(filename, modes[i]);
+
+		TEST_ASSERT_NOT_NULL_MESSAGE(f, tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT32_MESSAGE((int32_t)ftell(f), (int32_t)ftello(f), tellMode(modes[i]));
+
+		if (STR_EQUAL("w", modes[i]) || STR_EQUAL("wb", modes[i])) { /* can't read in "w" mode */
+			TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+			TEST_ASSERT_EQUAL_MESSAGE(0, remove(filename), tellMode(modes[i]));
+			continue;
+		}
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fseek(f, 0, SEEK_SET), tellMode(modes[i]));
+
+		{
+			int i = 1;
+			int errnoBefore;
+			char c;
+			while ((c = fgetc(f)) != -1) {
```

Comment:
On our ARM architectures targets char is from 0 to 255,, maybe use signed char or int for `c`